vispy/visuals/colorbar.py

```diff
@@ -436,76 +314,125 @@ def _update(self):
+        (label_pos, ticks_pos) = \
+            ColorBarVisual._calc_positions(center=self._center_pos,
+                                           halfdim=self._halfdim,
+                                           border_width=self.border_width,
+                                           orientation=self._orientation,
+                                           transforms=self.transforms)
 
-        elif self._orientation == "top":
-            text_x = x
-            text_y = y - _TEXT_PADDING_FACTOR * (halfh + self.border_width)
-            self._label.pos = text_x, text_y
+        self._label.pos = label_pos
+        self._ticks[0].pos = ticks_pos[0]
+        self._ticks[1].pos = ticks_pos[1]
 
-            self._ticks[0].pos = x - halfw, text_y
-            self._ticks[1].pos = x + halfw, text_y
-
-        elif self._orientation == "right":
-            text_x = x + _TEXT_PADDING_FACTOR * (halfw + self.border_width)
-            text_y = y
-            self._label.pos = text_x, text_y
+        if self._orientation == "right" or self._orientation == "left":
             self._label.rotation = -90
+            self._ticks[0].rotation = -90
+            self._ticks[1].rotation = -90
 
-            # TODO, HACK: See comment about ticks on "horizontal" conditional
-            self._ticks[0].pos = text_x, y + halfh
-            self._ticks[1].pos = text_x, y - halfh
-            self._ticks[0].rotation = self.ticks[1].rotation = -90
-
-        elif self._orientation == "left":
-            text_x = x - _TEXT_PADDING_FACTOR * (halfw + self.border_width)
-            text_y = y
-            self._label.pos = text_x, text_y
-            self._label.rotation = -90
-
-            self._ticks[0].pos = text_x, y + halfh
-            self._ticks[1].pos = text_x, y - halfh
-            self._ticks[0].rotation = self.ticks[1].rotation = -90
-
-        else:
-            # raise an error since the orientation is now what was
-            # expected
-            raise ColorBarVisual._get_orientation_error(self._orientation)
-
-    def _prepare_transforms(self, view):
-        print("preparing transorms")
-        CompoundVisual._prepare_transforms(view)
-        view._update()
-
-    @staticmethod
-    def _get_orientation_error(orientation):
-        return ValueError("orientation must"
-                          " be 'horizontal' or 'vertical', "
-                          "not '%s'" % (orientation, ))
+    def _prepare_draw(self, view):
+        # print ("preparing draw")
+        self._update_positions()
 
     @staticmethod
     def _get_anchors(orientation):
         if orientation == "top":
             return "center", "bottom"
+
         elif orientation == "bottom":
             return "center", "top"
+
         elif orientation == "left":
             return "center", "bottom"
+
         else:  # orientation == "right"
             return "center", "top"
 
+    @staticmethod
+    def _calc_positions(center, halfdim, border_width,
```

Comment:
All the ugly branching calculations in handled in one function, so yay, neater code :)